Create make more characters (resolves #227)

mkrause-nps opened the pull request:
Closes #227 
To test:

Run the Introduction scenario and observe that Moe and Curly wear shirts of different color.

mkrause-nps committed:
Added all male characters

mkrause-nps committed:
Added all male and female, with correct cloths

mkrause-nps committed:
Added logic to give each character user a different look

CowboyCoderForever commented:
It looks like you missed setting all the references after changing the character mesh in the new prefabs.

For example:
![image](https://user-images.githubusercontent.com/7308321/109325936-62a6bd00-780b-11eb-99de-12c30fc952b3.png)

There is also a reference in the Speech Bubble Root that you might have to ping Ryan about.
![image](https://user-images.githubusercontent.com/7308321/109326046-89fd8a00-780b-11eb-9d8a-dec6f35d9dab.png)

And I always quick-check my changes by playing a couple different scenarios. If you run the Starting Scenarios -> Introduction Scenarios its quite obvious there is something wrong with the prefabs (scrolling exceptions).

CowboyCoderForever commented:
Also, the idea of this feature is to have reproducible, incremental characters. For example, playing the Introduction Scenario five times in a row should produce the same two avatars each time for Moe and Curly. Moe gets avatar 1, Curly gets avatar 2, for example. 

What happens when you run this code multiple times in a row?

CowboyCoderForever commented on Assets/Code/Scriptable Variables/AvatarPrefabMapping.cs:
```diff
@@ -21,7 +21,9 @@ public class AvatarPrefabMapping : ScriptableVariable {
+      var mapping = _prefabs.Find(x => x._gender == gender && x._isTaken == false);
+      int idx = _prefabs.FindIndex(x => x._prefab.name == mapping._prefab.name);
+      _prefabs[idx]._isTaken = true;
       return mapping?._prefab;
```
What's going to happen here if all the prefabs are isTaken == true?  What we probably want is to continue cycling through the list. The scenario will have duplicate avatars in the scene, but that is better than no avatars.

There is actually a subtle issue here. You are modifying a ScriptableObject that auto-serializes -- it actually saves your runtime changes to file while Playing! You can see this by using the Unity Inspector to view the AvatarPrefabMapping .asset as you play. When you stop Playing, the .asset still contains your changes.

You can fix this by either:
1) don't modify the mapping scriptable object during runtime -- move your "isTaken" to outside the AvatarPrefabMapping class, or you leave it _inside_ the class and mark it with a `[NonSerialized]` attribute maybe.
2) "reset" the scriptableObject every time the game is started and ended. Maybe the `UserFactory` would call a reset function on the AvatarPrefabMapping during its `Awake()` and `OnDestroy()`

mkrause-nps committed:
Added animation and center of mass references to characters

mkrause-nps committed:
Prefab list resets at start, destroy, or when running out of prefabs

mkrause-nps committed:
Merge branch 'develop' into issue/227-make-more-characters

mkrause-nps committed:
Fixed case if .Find returns null

mkrause-nps committed:
Merge branch 'develop' into issue/227-make-more-characters

mkrause-nps committed:
Update of field _isTaken is simplified

CowboyCoderForever committed:
-Removed some duplicate components
-Fixed incorrect references for the AnimationFreezer
-Renamed a sub-folder for clarity.

CowboyCoderForever committed:
Reset some in-game values of this Scriptable Variable.
